core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.encoding
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+
+
+/**
+ * Encodes composite elements in a specific order managed by [mapElementIndex].
+ *
+ * This encoder will replicate the behavior of a standard encoding, but calling the `encode*Element` methods in
+ * the order defined by [mapElementIndex]. It first buffers each `encode*Element` calls in an array following
+ * the given indexes using [mapElementIndex], then when [endStructure] is called, it encodes the buffered calls
+ * in the expected order by replaying the previous calls on a concrete [CompositeEncoder] provided
+ * by [beginStructure].
+ *
+ * @param structureDescriptor descriptor of the structure being encoded
+ * @param beginStructure provides the [CompositeEncoder] to be used to encode the given descriptor's elements in the expected order.
```

Comment:
This parameter is quite surprising/confusing and can probably be documented more (or renamed). It is used as `beginStructure` in this type's `endStructure`, but from the caller perspective what it does is provide the delegate `CompositeEncoder` that will receive the reordered events.

Reply:
Replaced by `compositeEncoderDelegate: CompositeEncoder`